ecosystem/gstreamer_plugin/gst_mtl_st40p_tx.c

```diff
@@ -0,0 +1,426 @@
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+static gboolean gst_mtl_st40p_tx_sink_event(GstPad* pad, GstObject* parent,
+                                            GstEvent* event) {
+  Gst_Mtl_St40p_Tx* sink;
+  gint ret;
+
+  sink = GST_MTL_ST40P_TX(parent);
+
+  GST_LOG_OBJECT(sink, "Received %s event: %" GST_PTR_FORMAT, GST_EVENT_TYPE_NAME(event),
+                 event);
+
+  ret = GST_EVENT_TYPE(event);
+
+  switch (GST_EVENT_TYPE(event)) {
+    case GST_EVENT_SEGMENT:
+      if (!sink->tx_handle) {
+        GST_ERROR("Tx handle not initialized");
+        return FALSE;
+      }
+      ret = gst_pad_event_default(pad, parent, event);
+      break;
+    case GST_EVENT_EOS:
+      ret = gst_pad_event_default(pad, parent, event);
+      gst_element_post_message(GST_ELEMENT(sink), gst_message_new_eos(GST_OBJECT(sink)));
+      break;
+    default:
+      ret = gst_pad_event_default(pad, parent, event);
+      break;
+  }
+
+  return ret;
+}
+
+static void fill_st40_meta(struct st40_frame* frame, void* data, guint32 data_size) {
+  frame->meta[0].c = 0;
+  frame->meta[0].line_number = 10;
+  frame->meta[0].hori_offset = 0;
+  frame->meta[0].s = 0;
+  frame->meta[0].stream_num = 0;
+  frame->meta[0].did = 0x43;   // TODO: parametrize
+  frame->meta[0].sdid = 0x02;  // TODO: parametrize
+  frame->meta[0].udw_size = data_size;
+  frame->meta[0].udw_offset = 0;
+  frame->data = data;
+  frame->meta_num = 1;
+}
+
+/*
+ * Takes the buffer from the source pad and sends it to the mtl library via
+ * frame buffers, supports incomplete frames. But buffers needs to add up to the
+ * actual frame size.
+ */
+static GstFlowReturn gst_mtl_st40p_tx_chain(GstPad* pad, GstObject* parent,
+                                            GstBuffer* buf) {
+  Gst_Mtl_St40p_Tx* sink = GST_MTL_ST40P_TX(parent);
+  gint buffer_n = gst_buffer_n_memory(buf);
+  struct st40_frame_info* frame = NULL;
+  GstMemory* gst_buffer_memory;
+  GstMapInfo map_info;
+  gint bytes_to_write, bytes_to_write_cur;
+  void* cur_addr_buf;
+
+  if (!sink->tx_handle) {
+    GST_ERROR("Tx handle not initialized");
+    return GST_FLOW_ERROR;
+  }
+
+  for (int i = 0; i < buffer_n; i++) {
+    bytes_to_write = gst_buffer_get_size(buf);
+    gst_buffer_memory = gst_buffer_peek_memory(buf, i);
+
+    if (!gst_memory_map(gst_buffer_memory, &map_info, GST_MAP_READ)) {
+      GST_ERROR("Failed to map memory");
+      return GST_FLOW_ERROR;
+    }
+
+    while (bytes_to_write > 0) {
+      frame = st40p_tx_get_frame(sink->tx_handle);
+      if (!frame) {
+        GST_ERROR("Failed to get frame");
+        return GST_FLOW_ERROR;
+      }
+      cur_addr_buf = map_info.data + gst_buffer_get_size(buf) - bytes_to_write;
+      bytes_to_write_cur =
+          bytes_to_write > sink->frame_size ? sink->frame_size : bytes_to_write;
+      mtl_memcpy(frame->udw_buff_addr, cur_addr_buf, bytes_to_write_cur);
+      fill_st40_meta(frame->anc_frame, frame->udw_buff_addr, bytes_to_write_cur);
```

Comment:
there is no check for anc_frame.meta being null on the application side 

I don't mind IMO if we have certainty its not its fine but meaby lets make the function inline then ?
as a function with purpose we could also add the check for null ptr to make sure mby ? 


this is just a suggestion i am totally comfortable with letting it go as is just the function looks scary a lil bit

Reply:
meta, as dot in anc_frame.meta suggests, is allocated alongside `struct st40_frame` in which it is embedded.

anc_frame could theoretically be null though. I will add the check.

Reply:
After a thinking or just added comment to the function. It is quite normal to split function into multiple static function, and only this was my intention. Compiler probably will inline it anyway.

About the checks. When we get the frame from get_frame() we always assume that buffers are not NULL, the same in this case. I added only one check in library to make sure that it is not null.